pkg/ingress/client.go

```diff
@@ -17,43 +17,45 @@ import (
 
 // NewIngressClient implements ingress.Monitor and creates the Kubernetes client to monitor Ingress resources.
 func NewIngressClient(kubeClient kubernetes.Interface, kubeController k8s.Controller, stop chan struct{}, cfg configurator.Configurator) (Monitor, error) {
-	// TODO(#2798): Dynamically retrieve configured version
-	// Currently, since only networking.k8s.io/v1beta1 is supported, hardcode this.
-	requestedAPIVersion := IngressNetworkingV1beta1
+	k8sServerVersion, err := k8s.GetKubernetesServerVersionNumber(kubeClient)
+	if err != nil {
+		log.Error().Err(err).Msgf("Error retrieving k8s server version required to initialize ingress client")
+		return nil, err
+	}
 
-	var informer cache.SharedIndexInformer
+	// Initialize the version specific ingress informers and caches
 	informerFactory := informers.NewSharedInformerFactory(kubeClient, k8s.DefaultKubeEventResyncInterval)
+	ingressEventTypes := k8s.EventTypes{
+		Add:    announcements.IngressAdded,
+		Update: announcements.IngressUpdated,
+		Delete: announcements.IngressDeleted,
+	}
 
-	switch requestedAPIVersion {
-	case IngressNetworkingV1:
-		informer = informerFactory.Networking().V1().Ingresses().Informer()
-
-	case IngressNetworkingV1beta1:
-		informer = informerFactory.Networking().V1beta1().Ingresses().Informer()
-
-	default:
-		return nil, ErrUnsupportedAPIVersion
+	shouldObserve := func(obj interface{}) bool {
+		ns := reflect.ValueOf(obj).Elem().FieldByName("ObjectMeta").FieldByName("Namespace").String()
+		return kubeController.IsMonitoredNamespace(ns)
 	}
 
 	client := Client{
-		informer:       informer,
-		cache:          informer.GetStore(),
 		cacheSynced:    make(chan interface{}),
 		kubeController: kubeController,
-		apiVersion:     requestedAPIVersion,
 	}
 
-	shouldObserve := func(obj interface{}) bool {
-		ns := reflect.ValueOf(obj).Elem().FieldByName("ObjectMeta").FieldByName("Namespace").String()
-		return kubeController.IsMonitoredNamespace(ns)
+	// If k8s server version >= 1.19, initialize the v1 informer
+	// Ref: https://kubernetes.io/docs/reference/using-api/deprecation-guide/#ingress-v122
+	if k8sServerVersion[0] >= 1 && k8sServerVersion[1] >= 19 {
```

Comment:
This is good and makes sense.

I wonder if we should instead use the Go client's alternative of:
```console
$ kubectl explain ingress

KIND:     Ingress
VERSION:  extensions/v1beta1
...
```
and this could instruct what Ingress version we should use.

Then we would not have to encode Kubernetes versions.

@shashankram what do you think?

Reply:
@draychev It is possible to figure out dynamically the API groups the k8s server supports, through go-client's discovery pkg. It would imply walking all resource kinds that belong to the `networking.k8s.io` API group, and matching `Kind == Ingress`.

I'll explore this option and consider swapping the versioning checks in favor of negotiating the supported API versions with the k8s API server. I am inclined to address this in a follow-up though.

Reply:
@draychev https://github.com/openservicemesh/osm/pull/3101 should address your suggestion.